layers/ipfix.go

```diff
@@ -0,0 +1,773 @@
+//
+//     0                   1                   2                   3
+//     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |   Set ID = Template ID        |          Length               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |   Record 1 - Field Value 1    |   Record 1 - Field Value 2    |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |   Record 1 - Field Value 3    |             ...               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |   Record 2 - Field Value 1    |   Record 2 - Field Value 2    |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |   Record 2 - Field Value 3    |             ...               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |   Record 3 - Field Value 1    |   Record 3 - Field Value 2    |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |   Record 3 - Field Value 3    |             ...               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |              ...              |      Padding (optional)       |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+type IPFIXDataSet struct {
+	// Set header
+	Header IPFIXSetHeader
+	// Data records in the set
+	Records []IPFIXDataRecord
+	// Raw bytes
+	Bytes []byte
+}
+
+// Len returns the content length in octet of a Data Set
+func (ds IPFIXDataSet) Len() int {
+	return int(ds.Header.Length)
+}
+
+func (ds *IPFIXDataSet) decodeFromBytes(data *[]byte, t IPFIXTemplateAccessor) error {
+	if len(*data) < ds.Len()-ds.Header.Len() {
+		*data = []byte{}
+		return fmt.Errorf("invalid data set length")
+	}
+
+	dataLen := ds.Len() - ds.Header.Len()
+	ds.Bytes = (*data)[:dataLen]
+
+	if t != nil {
+		// a Data Set can contains multiple Data Records
+		// we cannot know the exact length of a data record due to the variable
+		// length of certain field but we know the minimum. So until we have
+		// enough data for a record and we did not exceed the data set length,
+		// we can decode another record.
+		dataRead := 0
+		for len(*data) >= int(t.GetMinDataRecordLength()) && dataRead < dataLen {
+			dr := &IPFIXDataRecord{
+				ID:     ds.Header.ID,
+				Fields: []IPFIXField{},
+			}
+			n, err := dr.decodeFromBytes(data, t)
+			if err != nil {
+				return err
+			}
+			ds.Records = append(ds.Records, *dr)
+			dataRead += n
+		}
+
+		// remaining octet corresponding to padding see RFC 7011 section 3.1.1
+		padding := dataLen - dataRead
+		if padding > 0 {
+			for _, b := range (*data)[:padding] {
+				if b != 0x00 {
+					return fmt.Errorf("invalid data set padding value")
+				}
+			}
+			*data = (*data)[padding:]
+		}
+	} else {
+		*data = (*data)[dataLen:]
```

Comment:
Just curious how does collector use this data without any template? Does it buffer and wait for the template?

Reply:
yes that is the RFC recommendation https://datatracker.ietf.org/doc/html/rfc7011#section-9.3